apps/language_server/lib/language_server/providers/code_lens.ex

```diff
@@ -110,30 +112,96 @@ defmodule ElixirLS.LanguageServer.Providers.CodeLens do
     end
   end
 
-  def code_lens(server_instance_id, uri, text) do
+  def spec_code_lens(server_instance_id, uri, text) do
     resp =
       for {_, line, {mod, fun, arity}, contract, is_macro} <- Server.suggest_contracts(uri),
           SourceFile.function_def_on_line?(text, line, fun),
           spec = ContractTranslator.translate_contract(fun, contract, is_macro) do
-        %{
-          "range" => range(line - 1, 0, line - 1, 0),
-          "command" => %{
-            "title" => "@spec #{spec}",
-            "command" => "spec:#{server_instance_id}",
-            "arguments" => [
-              %{
-                "uri" => uri,
-                "mod" => to_string(mod),
-                "fun" => to_string(fun),
-                "arity" => arity,
-                "spec" => spec,
-                "line" => line
-              }
-            ]
+        build_code_lens(
+          line,
+          "@spec #{spec}",
+          "spec:#{server_instance_id}",
+          %{
+            "uri" => uri,
+            "mod" => to_string(mod),
+            "fun" => to_string(fun),
+            "arity" => arity,
+            "spec" => spec,
+            "line" => line
           }
-        }
+        )
       end
 
     {:ok, resp}
   end
+
+  def test_code_lens(uri, src) do
+    file_path = SourceFile.path_from_uri(uri)
+
+    if imports?(src, ExUnit.Case) do
+      test_calls = calls_to(src, :test)
+      describe_calls = calls_to(src, :describe)
+
+      calls_lenses =
+        for {line, _col} <- test_calls ++ describe_calls do
+          test_filter = "#{file_path}:#{line}"
+
+          build_code_lens(line, "Run test", "elixir.test.run", test_filter)
+        end
+
+      file_lens = build_code_lens(1, "Run test", "elixir.test.run", file_path)
+
+      {:ok, [file_lens | calls_lenses]}
+    end
+  end
+
+  @spec imports?(String.t(), [atom()] | atom()) :: boolean()
+  defp imports?(buffer, modules) do
+    buffer_file_metadata =
+      buffer
+      |> Parser.parse_string(true, true, 1)
+
+    imports_set =
+      buffer_file_metadata.lines_to_env
+      |> get_imports()
+      |> MapSet.new()
+
+    modules
+    |> List.wrap()
+    |> MapSet.new()
+    |> MapSet.subset?(imports_set)
+  end
+
+  defp get_imports(lines_to_env) do
```

Comment:
This won't work for files with more than 1 module and basically with anything after the module. lines_to_env contains imports in the current scope/line (`k` is line). Not every line is registered currently but it may change in the future. I wouldn't count on last line in file containing any useful info.